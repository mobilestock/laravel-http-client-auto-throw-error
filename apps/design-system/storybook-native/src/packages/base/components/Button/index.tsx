import React, { PropsWithChildren } from 'react'
import { ActivityIndicator, StyleProp, TextStyle, TouchableHighlightProps } from 'react-native'
import styled, { css } from 'styled-components/native'

import { theme } from '../../utils/theme'

export interface PropsButton extends TouchableHighlightProps {
  text?: string
  isLoading?: boolean
  textStyle?: StyleProp<TextStyle>
}

export const Button: React.FC<PropsWithChildren<PropsButton>> = props => {
  return (
    <ButtonStyle {...props}>
      {props.isLoading ? (
        <ActivityIndicator color={theme.colors.decorator.pure} size={25} />
      ) : (
        <>{props.children ? props.children : <Text style={props.textStyle}>{props.text}</Text>}</>
      )}
    </ButtonStyle>
  )
}

const ButtonStyle = styled.TouchableHighlight`
  background-color: ${theme.colors.button.base};
  min-height: ${theme.layout.height(2)}px;
  justify-content: center;
  align-items: center;
  border-radius: ${theme.layout.size(1)}px;
  margin: ${theme.layout.size(1)}px;
  padding: ${theme.layout.size(1)}px;

  ${({ disabled }) =>
    disabled &&
    css`
      opacity: 0.5;
    `}
`

const Text = styled.Text`
  color: ${theme.colors.text.secondary};
  font-size: ${theme.fonts.size(3)}px;
  text-align: center;
`
