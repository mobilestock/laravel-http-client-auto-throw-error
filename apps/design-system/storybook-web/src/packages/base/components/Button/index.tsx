import React, { ButtonHTMLAttributes } from 'react'
import styled from 'styled-components'

import { CircularProgress } from '@mui/material'

import { theme } from '../../utils/theme'

export interface PropsButton extends ButtonHTMLAttributes<HTMLButtonElement> {
  text?: string
  isLoading?: boolean
}

export const Button: React.FC<PropsButton> = props => {
  return (
    <ButtonStyle {...props}>
      <span className="emphasis">
        {props.isLoading ? <CircularProgress className="circular" /> : <>{props.text}</>}
      </span>
      {props.isLoading || props.children}
    </ButtonStyle>
  )
}

const ButtonStyle = styled.button`
  display: flex;
  align-items: center;
  justify-content: center;

  width: 100%;
  min-height: 3rem;
  padding: 0.9rem 1rem;

  font-size: 0.9rem;
  font-family: 'Open Sans', sans-serif;
  font-weight: 400;

  cursor: pointer;

  box-shadow: 0 0.25rem 0.25rem rgba(0, 0, 0, 0.1);
  border: none;
  border-radius: 0.4rem;
  background-color: ${theme.colors.button.base};
  color: ${theme.colors.text.secondary};
  svg {
    font-size: 1.6rem;
    margin-right: 0.3rem;
  }
  &:hover {
    opacity: 0.8;
  }
  .emphasis {
    text-transform: uppercase;
  }
  .circular {
    width: 1rem !important;
    height: 1rem !important;
    color: ${theme.colors.text.secondary};
    svg {
      margin-right: 0;
    }
  }
  &:disabled {
    opacity: 0.7;
  }
`
