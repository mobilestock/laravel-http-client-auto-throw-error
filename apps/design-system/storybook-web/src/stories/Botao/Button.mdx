import { Canvas, Meta, Story } from '@storybook/blocks';
import * as ButtonStories from './index.stories';

<Meta of={ButtonStories} />

# Botão

O componente **Botão** é utilizado para ações gerais em formulários e interfaces, suportando variações como `Carregando`, `Desativado`, e diferentes estilos.

## Uso Básico

Você pode utilizar o botão com um texto padrão e sem estado de carregamento.

<Canvas of={ButtonStories.Primary} />

## Estados do Botão

### Carregando

Use a prop `isLoading` para indicar um estado de carregamento no botão.

<Canvas of={ButtonStories.Loading} />

### Desativado

Use a prop `disabled` para desativar o botão e impedir interações.

<Canvas of={ButtonStories.Disabled} />
