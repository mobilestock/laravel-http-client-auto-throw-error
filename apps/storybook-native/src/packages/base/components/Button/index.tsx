import React, { PropsWithChildren } from 'react';
import { ActivityIndicator, StyleProp, TextStyle, TouchableHighlightProps } from 'react-native';
import styled, { css } from 'styled-components/native';

import { theme } from '../../utils/theme';

export interface PropsButton extends TouchableHighlightProps {
  text?: string;
  isLoading?: boolean;
  textStyle?: StyleProp<TextStyle>;
}

export const Button: React.FC<PropsWithChildren<PropsButton>> = (props) => {
  return (
    <ButtonStyle {...props}>
      {props.isLoading ? (
        <ActivityIndicator color={theme.colors.container.default} size={25} />
      ) : (
        <>{props.children ? props.children : <Text style={props.textStyle}>{props.text}</Text>}</>
      )}
    </ButtonStyle>
  );
};

const ButtonStyle = styled.TouchableHighlight`
  background-color: ${({ theme }) => theme.colors.button.default};
  min-height: ${({ theme }) => theme.layout.height(2)}px;
  justify-content: center;
  align-items: center;
  border-radius: ${({ theme }) => theme.layout.size(2)}px;
  margin: ${({ theme }) => theme.layout.size(1)}px;
  padding: ${({ theme }) => theme.layout.size(2)}px;

  ${({ disabled }) =>
    disabled &&
    css`
      opacity: 0.5;
    `}
`;

const Text = styled.Text`
  color: ${({ theme }) => theme.colors.text.secondary};
  font-size: ${({ theme }) => theme.fonts.size(16)}px;
  text-align: center;
`;
