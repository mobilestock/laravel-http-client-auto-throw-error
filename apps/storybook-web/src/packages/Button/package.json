{
  "name": "@mobilestockweb/button",
  "version": "0.0.1",
  "main": "./src/index.tsx",
  "files": [
    "*"
  ],
  "exports": {
    ".": {
      "default": "./src/index.tsx"
    }
  },
  "dependencies": {
    "@mui/material": "^5.15.13"
  },
  "peerDependencies": {
    "react": "^18.2.0",
    "react-dom": "^18.2.0",
    "styled-components": "^6.1.8"
  }
}