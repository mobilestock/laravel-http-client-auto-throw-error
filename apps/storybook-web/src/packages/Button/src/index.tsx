import { CircularProgress } from '@mui/material'
import { ButtonHTMLAttributes } from 'react'
import styled from 'styled-components'

export interface PropsButton extends ButtonHTMLAttributes<HTMLButtonElement> {
  text?: string
  isLoading?: boolean
}

export default function Button({ isLoading, text, children, ...rest }: PropsButton) {
  return (
    <ButtonStyle {...rest}>
      {isLoading ? (
        <CircularProgress className="circular" />
      ) : (
        <span className="emphasis">{text || children}</span>
      )}
    </ButtonStyle>
  )
}

const ButtonStyle = styled.button`
  display: flex;
  align-items: center;
  justify-content: center;

  width: 100%;
  min-height: 3rem;
  padding: 0.9rem 1rem;

  font-size: 0.9rem;
  font-family: 'Open Sans', sans-serif;
  font-weight: 400;

  cursor: pointer;

  box-shadow: 0 0.25rem 0.25rem ${({ theme }) => theme.colors.button.shadow};
  border: none;
  border-radius: 0.4rem;
  background-color: ${({ theme }) => theme.colors.button.default};
  color: ${({ theme }) => theme.colors.text.default};
  svg {
    font-size: 1.6rem;
    margin-right: 0.3rem;
  }
  &:hover {
    opacity: 0.8;
  }
  .emphasis {
    text-transform: uppercase;
  }
  .circular {
    width: 1rem !important;
    height: 1rem !important;
    color: ${({ theme }) => theme.colors.text.default};
    svg {
      margin-right: 0;
    }
  }
  &:disabled {
    opacity: 0.7;
  }
`
