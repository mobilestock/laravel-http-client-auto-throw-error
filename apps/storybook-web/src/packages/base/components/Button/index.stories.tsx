import type { Meta, StoryObj } from '@storybook/react'

import { Button } from './index'

const meta = {
  title: 'Componentes/Button',
  component: Button,
  parameters: {
    layout: 'centered',
    docs: {
      subtitle: 'Botão padrão do sistema.'
    }
  },
  args: {
    isLoading: false,
    text: 'Botão',
    onClick: () => console.log('Clicou no botão')
  },
  argTypes: {
    isLoading: {
      control: 'boolean',
      description: 'Indica se o botão está em estado de carregamento.',
      defaultValue: false
    },
    text: {
      control: 'text',
      description: 'Texto exibido dentro do botão.',
      defaultValue: 'Botão'
    },
    onClick: {
      action: () => {
        alert('Clicou no botão')
      },
      description: 'Função chamada ao clicar no botão.',
      defaultValue: () => console.log('teste'),
      control: () => {
        alert('Clicou no botão')
      },
      type: { name: 'function' }
    }
  }
} satisfies Meta<typeof Button>

export default meta
type Story = StoryObj<typeof meta>

export const Primary: Story = {
  args: {
    isLoading: false,
    text: 'Botão'
  }
}

export const Loading: Story = {
  args: {
    style: { minWidth: '9rem' },
    isLoading: true,
    text: 'Carregando...',
    disabled: true
  }
}

export const Disabled: Story = {
  args: {
    isLoading: false,
    text: 'Desativado',
    disabled: true
  }
}
