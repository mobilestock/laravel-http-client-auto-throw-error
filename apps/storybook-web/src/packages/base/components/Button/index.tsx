import { CircularProgress } from '@mui/material'
import React, { ButtonHTMLAttributes } from 'react'
import styled from 'styled-components'

export interface PropsButton extends ButtonHTMLAttributes<HTMLButtonElement> {
  text?: string
  isLoading?: boolean
}

export const Button: React.FC<PropsButton> = (props) => {
  return (
    <ButtonStyle {...props}>
      <span className="emphasis">
        {props.isLoading ? <CircularProgress className="circular" /> : <>{props.text}</>}
      </span>
      {props.isLoading || props.children}
    </ButtonStyle>
  )
}

const ButtonStyle = styled.button`
  display: flex;
  align-items: center;
  justify-content: center;

  width: 100%;
  min-height: 3rem;
  padding: 0.9rem 1rem;

  font-size: 0.9rem;
  font-family: 'Open Sans', sans-serif;
  font-weight: 400;

  cursor: pointer;

  box-shadow: 0 0.25rem 0.25rem ${({ theme }) => theme.colors.button.shadow};
  border: none;
  border-radius: 0.4rem;
  background-color: ${({ theme }) => theme.colors.button.default};
  color: ${({ theme }) => theme.colors.text.default};
  svg {
    font-size: 1.6rem;
    margin-right: 0.3rem;
  }
  &:hover {
    opacity: 0.8;
  }
  .emphasis {
    text-transform: uppercase;
  }
  .circular {
    width: 1rem !important;
    height: 1rem !important;
    color: ${({ theme }) => theme.colors.text.default};
    svg {
      margin-right: 0;
    }
  }
  &:disabled {
    opacity: 0.7;
  }
`
