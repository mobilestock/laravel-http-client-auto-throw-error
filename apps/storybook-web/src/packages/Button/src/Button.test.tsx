import '@testing-library/jest-dom'
import { fireEvent, render } from '@testing-library/react'

import Button from '.'

describe('Button Component - Web', () => {
  it('deve renderizar sem erros', () => {
    render(global.app(<Button />))
  })

  it('deve exibir o texto passado via props', () => {
    const { getByText } = render(global.app(<Button text="Clique aqui" />))
    getByText('Clique aqui')
  })

  it('deve renderizar os filhos quando passados', () => {
    const { getByText } = render(global.app(<Button>Enviar</Button>))
    getByText('Enviar')
  })

  it('deve mostrar o indicador de carregamento quando isLoading for true', () => {
    const { container } = render(global.app(<Button isLoading={true} />))
    const spinner = container.querySelector('.circular')
    expect(spinner).toBeInTheDocument()
  })

  it('nÃ£o deve mostrar o indicador de carregamento quando isLoading for false', () => {
    const { container } = render(global.app(<Button isLoading={false} />))
    const spinner = container.querySelector('.circular')
    expect(spinner).toBeNull()
  })

  it('deve chamar onClick quando clicado', () => {
    const onClick = jest.fn()
    const { getByRole } = render(global.app(<Button onClick={onClick} />))
    fireEvent.click(getByRole('button'))
    expect(onClick).toHaveBeenCalled()
  })

  it('deve estar desabilitado quando a prop disabled for true', () => {
    const { getByRole } = render(global.app(<Button disabled={true} />))
    expect(getByRole('button')).toBeDisabled()
  })

  it('deve aplicar a classe personalizada quando className for fornecido', () => {
    const { getByRole } = render(global.app(<Button className="custom-class" />))
    expect(getByRole('button')).toHaveClass('custom-class')
  })
})
