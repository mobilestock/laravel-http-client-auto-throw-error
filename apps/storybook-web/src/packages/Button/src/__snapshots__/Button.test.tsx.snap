// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component - Web deve corresponder ao snapshot 1`] = `
<DocumentFragment>
  <button
    class="sc-blHHSb ehLeRL"
  >
    <span
      class="emphasis"
    />
  </button>
</DocumentFragment>
`;
